devtools/x/src/test.rs

```diff
@@ -92,49 +100,136 @@ pub fn run(mut args: Args, xctx: XContext) -> Result<()> {
         env: &env_vars,
     };
 
-    if args.unit {
+    let cmd_result = if args.unit {
         cmd.run_with_exclusions(
             config.system_tests().iter().map(|(p, _)| p),
             &CargoArgs::default(),
-        )?;
+        )
     } else if !args.package.is_empty() {
-        cmd.run_on_packages(args.package.iter(), &CargoArgs::default())?;
+        cmd.run_on_packages(args.package.iter(), &CargoArgs::default())
     } else if utils::project_is_root(&xctx)? {
         // TODO Maybe only run a subest of tests if we're not inside
         // a package but not at the project root (e.g. language)
-        cmd.run_on_all_packages(&CargoArgs::default())?;
+        cmd.run_on_all_packages(&CargoArgs::default())
     } else {
-        cmd.run_on_local_package(&CargoArgs::default())?;
+        cmd.run_on_local_package(&CargoArgs::default())
+    };
+
+    if !args.no_fail_fast && cmd_result.is_err() {
+        return cmd_result;
     }
 
-    if args.html_cov_dir.is_some() {
-        let debug_dir = project_root().join("target/debug/");
-        let mut grcov_html = Command::new("grcov");
-        grcov_html
-            //output file from coverage: gcda files
-            .arg(debug_dir.as_os_str())
-            //source code location
-            .arg("-s")
-            .arg(project_root().as_os_str())
-            //html output
-            .arg("-t")
-            .arg("html")
-            .arg("--llvm")
-            .arg("--branch")
-            .arg("--ignore")
-            .arg("/*")
-            .arg("--ignore")
-            .arg("x/*")
-            .arg("--ignore")
-            .arg("testsuite/*")
-            .arg("--ignore-not-existing")
-            .arg("-o")
-            .arg(args.html_cov_dir.unwrap());
-        info!("Build grcov Html Coverage Report");
-        info!("{:?}", grcov_html);
-        grcov_html.stdout(Stdio::inherit()).stderr(Stdio::inherit());
-        grcov_html.output()?;
+    if let Some(html_cov_dir) = &args.html_cov_dir {
+        create_dir_all(&html_cov_dir)?;
+        let html_cov_path = &html_cov_dir.canonicalize()?;
+        info!("created {}", &html_cov_path.to_string_lossy());
+        exec_grcov(&html_cov_path)?;
+    }
+    if let Some(html_lcov_dir) = &args.html_lcov_dir {
+        create_dir_all(&html_lcov_dir)?;
+        let html_lcov_path = &html_lcov_dir.canonicalize()?;
+        info!("created {}", &html_lcov_path.to_string_lossy());
+        exec_lcov(&html_lcov_path)?;
+        exec_lcov_genhtml(&html_lcov_path)?;
```

Comment:
Not sure if other people would get confused in which report to look at.  Maybe the grcov report is enough?  Will leave this you.